### ISSUE TITLE
Object.assign instead of manipulating input objects

### ISSUE THREAD

**Issue by SamVerschueren** [sindresorhus/eslint-plugin-unicorn#98]
I myself did it wrong in the early days, and I definitely come across this issue quite a lot.

```js
module.exports = opts => {
    opts.foo = opts.foo || 'bar';
};
```

While it should've been

```js
module.exports = opts => {
    opts = Object.assign({
        foo: 'bar'
    }, opts);
};
```

Not sure how easy it would be to detect this though.

**Comment by sindresorhus** [comment]
Slightly relevant issue: https://github.com/sindresorhus/eslint-config-xo/issues/18

**Comment by sindresorhus** [comment]
Seems like it should be possible to detect the arguments and to find the cases where something is assigning to properties of those arguments. Could maybe use [code path analysis](http://eslint.org/docs/developer-guide/code-path-analysis).

**Comment by MrHen** [comment]
Seems like the better fix is to make sure that objects in parameters are treated as immutable. The implementation for [https://github.com/jhusain/eslint-plugin-immutable#no-mutation](immutable/no-mutation) might work as a starting point.

```
// objects
function foo(bar) {
    del bar.zaz; // fail
    bar.zaz = bar.zaz || 'zaz'; // fail
    bar = {
        zaz: 'zaz',
        ...bar
    }; // pass
}

// arrays
function foo(bar) {
    bar.push('zaz'); // fail
    bar.pop() // fail
    bar.unshift('zaz'); // fail
    bar.shift(); // fail
    bar.length = 0; // fail
    delete bar[0]; // fail
    bar = {
        ...bar,
        'zaz'
    }; // pass
}
```

I'd argue that overshadowing parameter is also bad but should be a different rule.

**Comment by sindresorhus** [comment]
> There is also a functional difference between foo || bar and Object.assign. If you set a property to null and want a default then you can use || but not Object.assign.

I've been using `Object.assign` for options arguments for years and that has never been a problem in practice.

**Comment by MrHen** [comment]
It's not really a problem. It's just a slight functional difference. In my opinion, I think the `Object.assign` pattern is more clear / less surprising anyway.

In any case, I was just pondering whether I wanted to help out with this feature -- but in my case I'd rather have immutable parameters and/or use destructuring directly in the function signature. I'll go dig up something else to poke at.

**Comment by fregante** [comment]
Duplicate of https://github.com/sindresorhus/eslint-plugin-unicorn/issues/29
